fix:message list tool invocations

CalebBarnes opened the pull request:
## Summary

Fixes incorrect tool invocation format in message list that was causing client tools to fail during message format conversions.

## Problem

Tool invocations in the message list were not properly nested under a `toolInvocation` property when converting between AI SDK UI V5 Messages and Mastra V2 Messages. This caused the structure to be flat instead of nested:

**Before (incorrect):**
```typescript
{
  type: 'tool-invocation',
  toolCallId: '...',
  toolName: '...',
  args: {},
  result: {},
  state: 'result'
}
```

**After (correct):**
```typescript
{
  type: 'tool-invocation',
  toolInvocation: {
    toolCallId: '...',
    toolName: '...',
    args: {},
    result: {},
    state: 'result'
  }
}
```

## Changes

### Message Conversion (index.ts:2408-2434)

- **Fixed `convertV5ToolUIPartToV2ToolInvocationPart`**: Now properly nests tool invocation data under `toolInvocation` property
- Added proper type safety using `satisfies ToolInvocationUIPart`
- Correctly handles both `output-available` and `call` states

### AI SDK V5 Conversion (index.ts:2181-2203)

- **Fixed `toAISDKV5UIMessages`**: Accesses tool invocation properties correctly through `part.toolInvocation` instead of casting
- Removed incorrect type casting that was masking the structural issue
- Added proper type annotations with `satisfies AIV5Type.ToolUIPart`

### Message Filtering (index.ts:2809-2828)

- **Improved `toModelMessages`**: Now filters out `input-available` state tool parts which aren't supported by `convertToModelMessages`
- Simplified filtering logic to only keep tool parts with output states (`output-available` or `output-error`)
- Added clearer comments explaining filtering behavior

### Test Improvements

- Added test verification for clipboard tool invocations in memory
- Fixed database path handling for bundled output directory
- Improved `onToolCall` handler to use `addToolResult` instead of returning values
- Added `sendAutomaticallyWhen` for proper automatic tool call sending

## Impact

This fix ensures that:
- Client-side tools work correctly with the message list
- Message format conversions between AI SDK formats and Mastra formats maintain correct structure
- Tool invocations are properly typed and verified at compile time
- Memory persistence correctly stores and retrieves tool invocation data

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed tool-invocation formatting so client tools work correctly and tool-call states transition reliably (streaming → input/output).
  * Sanitization now preserves only output-bearing tool parts for model-facing requests.

* **Tests**
  * Updated integration and unit tests to reflect new memory message shape and tool-invocation handling, including streaming-to-result behavior.

* **Chores**
  * Added a changeset entry for a patch release.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Fixes tool-invocation message formatting and v5 UI message sanitization, updating conversions, tests, and integration routes to ensure client-side tools and memory work reliably.
> 
> - **Core (message-list)**:
>   - Correct conversion between AIV5 UI and Mastra V2: tool invocations are now nested under `toolInvocation` with proper types (`ToolInvocationUIPart`).
>   - Skip invalid `tool-invocation` parts without `toolInvocation` and other stray `tool-` parts when mapping.
>   - Convert AIV5 tool parts back to V2 using structured `toolInvocation` objects.
>   - Improve sanitization: filter out `input-available`/streaming parts, keep only `output-available`/`output-error`, and normalize `output.value`.
> - **Tests**:
>   - Update expected request payloads for single/multi-turn and memory to explicit message arrays (v1/v2).
>   - Adjust tool-call ordering assertions to expect `function_call` + `function_call_output` only.
> - **Integration (memory v5)**:
>   - Add server `chatRoute` at `/chat`; update `DefaultChatTransport` to use it.
>   - Enable client-side tools with `sendAutomaticallyWhen` and `addToolResult`; verify clipboard tool invocations persist in memory.
>   - Use dynamic DB path for bundled output and randomized `resourceId`.
> - **Release**:
>   - Patch bumps: `@mastra/core`, `@mastra/memory`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 07f92c15433b26257c43b9d915b46d6054e56368. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

CalebBarnes committed:
toolInvocation parts fix type and wrap with toolInvocation

CalebBarnes committed:
update usechat test

CalebBarnes committed:
Merge branch 'main' into fix/message-list-tool-invocations

CalebBarnes committed:
remove console log

CalebBarnes committed:
remove comment

CalebBarnes committed:
remove infinity timeout

CalebBarnes commented on packages/memory/integration-tests-v5/src/streaming-memory.test.ts:
```diff
@@ -253,15 +253,14 @@ describe('Memory Streaming Tests', () => {
         threadId,
         resourceId,
       });
-      await weatherAgent.generate(`LA weather`, { threadId, resourceId });
```
I don't think we expect anything from it anyways

CalebBarnes committed:
lint fix import order

CalebBarnes committed:
add changeset

CalebBarnes commented on packages/core/src/agent/message-list/index.ts:
```diff
@@ -2126,25 +2127,24 @@ export class MessageList {
           const inv = part.toolInvocation;
-          const v5State = inv.state === 'result' ? 'output-available' : 'input-available';
 
           if (inv.state === 'result' && inv.result) {
```
@TylerBarnes If the tool intentionally returns null, could we end up setting the state to `input-available` accidentally?

TylerBarnes replied:
yeah we probably dont need the `inv.result` check at all

NikAiyer committed:
Merge branch 'main' into fix/message-list-tool-invocations

TylerBarnes committed:
test: update stream test expectations for sanitizeV5UIMessages filtering

The sanitizeV5UIMessages function now filters out input-available tool parts
before passing to convertToModelMessages (which doesn't support them). This means
only output-available tool parts are converted, which generates one function_call
message per tool invocation instead of two.

Updated test expectations in stream.test.ts to reflect this behavior.

NikAiyer committed:
Merge branch 'main' into fix/message-list-tool-invocations

TylerBarnes committed:
fix: handle falsy tool invocation results correctly

Remove && inv.result check to correctly handle falsy results like false, 0, empty string.
Previously these valid results would be incorrectly filtered out.

TylerBarnes committed:
Merge branch 'main' into fix-message-list-tool-invocations

Resolved merge conflicts:
- packages/core/src/agent/__tests__/stream.test.ts: Kept PR's updated test expectations (1 function_call instead of 2)
- packages/core/src/agent/message-list/index.ts: Merged both imports (ToolInvocationUIPart from PR, randomUUID from main)
- packages/memory/integration-tests-v5/src/streaming-memory.test.ts: Merged imports and removed unused UIMessage type

All conflicts resolved in favor of preserving PR changes while incorporating main's updates from:
- PR #9255: Memory return format changes (MastraMessageV2 -> MastraDBMessage)
- PR #9588: Vendor AI-sdk v4 (randomUUID from @lukeed/uuid)

TylerBarnes committed:
fix(core): update stream.test.ts for memory API changes from main

- Fixed rememberMessages mock to return MastraDBMessage[] instead of AI SDK v4 Core messages
  - AI SDK v4 Core messages don't include threadId property needed by agent.ts filtering logic
  - Changed from list.get.remembered.aiV4.core() to list.get.remembered.db()
- Updated getMessages calls to destructure result.messages and remove deprecated format parameter
- Fixed list.get.remembered.v2() to list.get.remembered.db() for consistency with PR #9255

This resolves Core Package Tests failures after merging main.

TylerBarnes committed:
chore: trigger CI

TylerBarnes committed:
Merge branch 'main' into fix/message-list-tool-invocations

Resolved conflicts:
- packages/core/src/agent/message-list/index.ts: Combined PR's ToolInvocationUIPart import with main's vendored AI SDK v4 imports (@internal/ai-sdk-v4/message, @internal/ai-sdk-v4/tool) from PR #9594
- packages/core/src/agent/agent.ts: Removed debug console.log statement
- packages/core/src/processors/processors/structured-output.test.ts: Removed .only from focused test

Linting and build passed for @mastra/core after merge.

TylerBarnes committed:
fix: change callProviderMetadata to providerMetadata in ToolUIPart

Per AI SDK v5 spec, ToolUIPart uses 'providerMetadata' not 'callProviderMetadata'.
Updated both result and input-available state branches.

Addresses: https://github.com/mastra-ai/mastra/pull/9590#discussion_r1887234567

TylerBarnes committed:
fix: remove duplicate tool-invocation conversion block

The duplicate block at lines 2191-2213 was unreachable due to an earlier continue statement.
Keeping only the original implementation to avoid redundancy.

Addresses: https://github.com/mastra-ai/mastra/pull/9590#discussion_r1887234568

TylerBarnes committed:
fix: change .messagesV2 to .messages in streaming-memory.test.ts

The query() method returns { messages: MastraDBMessage[] }, not messagesV2.
Updated to access the correct property name.

Addresses: https://github.com/mastra-ai/mastra/pull/9590#discussion_r1887234569

TylerBarnes committed:
revert: keep callProviderMetadata instead of providerMetadata

The CodeRabbit suggestion to rename callProviderMetadata to providerMetadata was incorrect.
The type definitions (AIV5Type.ToolUIPart) expect callProviderMetadata, not providerMetadata.
Reverting commit 579cee5afc to fix TypeScript build errors.

Related: https://github.com/mastra-ai/mastra/pull/9590#discussion_r1882063253

TylerBarnes committed:
fix(core): skip tool-invocation parts without toolInvocation object

The previous fix to remove duplicate tool-invocation handling inadvertently
caused tool-invocation parts without a nested toolInvocation object to fall
through to the else block, causing a TypeScript error.

This fix explicitly skips tool-invocation parts that don't have a toolInvocation
object, preventing them from being pushed to the parts array.

Related to PR comment: https://github.com/mastra-ai/mastra/pull/9590#discussion_r1883488295

TylerBarnes committed:
Merge branch 'main' into fix/message-list-tool-invocations

DanielSLew committed:
merge main

DanielSLew committed:
fix test

DanielSLew committed:
fix lint

DanielSLew committed:
fix route

DanielSLew committed:
merge main

DanielSLew committed:
fix lint

DanielSLew committed:
Merge branch 'main' of github.com:mastra-ai/mastra into fix/message-list-tool-invocations

DanielSLew committed:
update route

DanielSLew committed:
fix tests

DanielSLew committed:
fix quote

DanielSLew committed:
Merge branch 'main' into fix/message-list-tool-invocations

DanielSLew committed:
revert kitchen sink

DanielSLew committed:
Merge branch 'fix/message-list-tool-invocations' of github.com:mastra-ai/mastra into fix/message-list-tool-invocations

DanielSLew committed:
Merge branch 'main' into fix/message-list-tool-invocations

DanielSLew committed:
Merge branch 'main' into fix/message-list-tool-invocations

DanielSLew committed:
Merge branch 'main' into fix/message-list-tool-invocations

DanielSLew committed:
Merge branch 'main' into fix/message-list-tool-invocations

DanielSLew committed:
fix test
